fix: support callv and proxies for lsp.js

ajbt200128 opened the pull request:
Previously we added support for proxies in #9994. But it broke LSP.js networking as we were forced to use callv to support proxies as a workaround, and callv isn't supported in cohttp-jsoo!. So we have to patch callv and patch our XHR js library to enable proxies for this stuff. I'm going to try and make a bunch of PRs to cohttp and ca-certs etc to fix this upstream, but for now we have this fix.

## Test plan

### For testing js w/out proxy
```bash
make build-semgrep-jsoo-debug
make -C js build
make -C js test
```
### For testing JS + proxy
```bash
make build-semgrep-jsoo-debug
make -C js build
make -C js/language_server
# copy js/language_server/dist into extension
# Run vscode with proxy and Use JS to confirm it works
```

TODO add MIT license to VSCode

ajbt200128 committed:
fix: support callv and proxies for lsp.js

Previously we added support for proxies in #9994. But it broke LSP.js
networking as we were forced to use callv to support proxies as a
workaround, and callv isn't supported in cohttp-jsoo!. So we have to
patch callv and patch our XHR js library to enable proxies for this
stuff. I'm going to try and make a bunch of PRs to cohttp and ca-certs
etc to fix this upstream, but for now we have this fix

kopecs commented on js/node_shared/XMLHttpRequest.js:
```diff
@@ -385,6 +385,17 @@ XMLHttpRequest = function () {
     }
 
+    // get if HTTP_PROXY or HTTPS_PROXY is set
+    var proxy = process.env[ssl ? "HTTPS_PROXY" : "HTTP_PROXY"];
```
```suggestion
    const proxy = process.env[ssl ? "HTTPS_PROXY" : "HTTP_PROXY"];
```

ajbt200128 committed:
add test for e2e making REAL network requests

ajbt200128 committed:
const

ajbt200128 committed:
fix tests

normal shared cant have cohttp in it because then it expects xhr.
There's xhr in the browser so it'd be weird to include it regardless of node

kopecs commented on js/node_shared/Semgrep_js_node_shared.ml:
```diff
@@ -0,0 +1,45 @@
+  let callv ?ctx (uri : Uri.t) stream =
+    (* Differs from request uri if HTTP(S)_PROXY is set. But we handle this in
+       the node_shared XHR*)
+    ignore uri;
```
```suggestion
```
Dead now?

kopecs commented on js/node_shared/XMLHttpRequest.js:
```diff
@@ -13,6 +13,26 @@
  * @author Dan DeFelippi <dan@driverdan.com>
  * @contributor David Ellis <d.f.ellis@ieee.org>
  * @license MIT
```
Is this whole file copied, or is it modified a fair amount? I would recommend having a note so we know what's what (like precisely what is just inbound versus derived versus ours and unrelated -- best of the last isn't in the same file but might make sense)

kopecs commented on src/osemgrep/language_server/Test_LS_e2e.ml:
```diff
@@ -664,6 +669,9 @@ let check_startup info folders (files : Fpath.t list) =
   let%lwt () = assert_notif rulesRefreshedNotif "semgrep/rulesRefreshed" in
+  let rules_length = List.length info.server.session.cached_session.rules in
+  Alcotest.(check bool)
+    "rules refreshed greater than 1" (rules_length > 20) true;
```
This is a test so it's probably fine but fyi we have a function to check if a list is shorter/longer than a fixed length if we don't need to count every element.

kopecs commented on js/language_server/dune:
```diff
@@ -19,13 +19,11 @@
  (modes js)
  (js_of_ocaml
   (javascript_files
    ;; Custom XMLHttpRequest implementation that doesn't preset the User-Agent header
```
```suggestion
```
Should also be removed given the deletion on the next line?

kopecs commented on js/node_shared/Semgrep_js_node_shared.ml:
```diff
@@ -0,0 +1,45 @@
+ *)
+
+(* Commentary *)
+(*  *)
```
A longer comment here explaining the use of `callv` (and anything else we later add) would be nice.

kopecs reviewed (approved):
Seems mostly good. I think we would benefit from some more documentation.

ajbt200128 committed:
fix/cleanup tests

ajbt200128 committed:
cleanup and comments

ajbt200128 committed:
remove thing

ajbt200128 committed:
oops

ajbt200128 committed:
merge develop

ajbt200128 committed:
I'm bad at merging

ajbt200128 commented:
pro checks failing seem unrelated
